design/task-management-api.md

```diff
@@ -0,0 +1,346 @@
+task may only be claimed by a call to `OwnTasks` if:
+* Its action is specified as acceptable to that call.
+* All dependencies of the task have been settled: all tasks specifying its task ID in their
+  `ToSignal` list have completed.
+* The task is not claimed by another worker.  Either:
+  - the task is in state `pending`, or
+  - the task is in state `in-progress`, but its `ActionDeadline` has elapsed (see "ownership
+    expiry", below).
+
+`OwnTasks` returns task IDs and a "performance token" for this performance of the task.
+Both ID and token must be provided to _return_ the task from ownership.  (The
+performance token is used to resolve conflicts during "ownership expiry", below.)
+
+Once a worker owns a task, it performs it.  It can decide to return the task to the task
+queue and _complete_, _abort_ or _retry_ it by calling `ReturnTask`.  Once completed, all
+dependents of the task are signalled, causing any dependent that has received all its
+required signals to be eligible for return by `OwnTasks`.
+
+#### Ownership expiry
+
+Processes can fail.  To allow restarting a failed process calls to `OwnTasks` may specify a
+deadline.  The lease granted to an owning worker will expire after this deadline, allowing
+another worker to own the task.  Only the _last_ worker granted ownership may call
+`ReturnTask` on the task.  A delayed worker should still return the task, in case the task
+has not yet been granted to another worker.
+
+#### Basic API
+
+This is a sample API.  All details are fully subject to change, of course!  Note that most
+`func`s are probably going to be methods on some object, which we assume will carry DB
+connection information etc.
+
+##### TaskData
+
+```go
+type TaskId string
+
+type ActorId string
+
+type PerformanceToken pgtype.UUID // With added stringifiers
+
+// TaskData describes a task to perform.
+type TaskData struct {
+	Id         TaskId              // Unique ID of task
+	Action     string              // Action to perform, used to fetch in OwnTasks
+	Body       *string             // Body containing details of action, used by clients only
+	Status     *string             // Human- and client-readable status
+	StatusCode TaskStatusCodeValue // Status code, used by task queue
+	NumTries   int                 // Number of times this task has moved from started to in-progress
+	MaxTries   *int                // Maximal number of times to try this task
+	// Dependencies might be stored or handled differently, depending on what gives reasonable
+	// performance.
+	TotalDependencies *int              // Number of tasks which must signal before this task can be owned
+	ToSignal          []TaskId          // Tasks to signal after this task is done
+	ActorId           ActorId           // ID of current actor performing this task (if in-progress)
+	ActionDeadline    *time.Time        // Deadline for current actor to finish performing this task (if in-progress)
```

Comment:
I guess this is the only locking mechanism possible with this design. But it is cumbersome. I based my design on DB advisory locks. It is more efficient, simpler and easier to use

Reply:
DB advisory locks assume a particular failure model, so they work for _some_ but not _all_ worker failures.  In particular, they use TCP keepalives, which are a very strange mechanism: the TCP/IP stack handles them, but by default on Linux it disconnects a TCP stream after 9 consecutive keepalive failures, each spaced 15 seconds apart, and starting only after 7200 seconds of quiet on the connection.

With the default PostgreSQL settings which are the default Linux settings (AFAICS we never change them, and given that they appear to be  global to the instance not all users will be able to change them), a VM failure will time out advisory locks after 7200 + 9*15 seconds == 7335 seconds, or 2 hours 2 minutes and 15 seconds (just 2 hours if the VM reboots to the same address, though).  (To _really_ cause trouble, suspend the process, too, and then it would _never_ time out because the TCP/IP stack continues to reply to keepalives...).

Ignore the suspense issue, if you set keepalives to more sane values then I end up with the same situation as here, and I _anyway_ need to identify and handle spurious timeouts.